R/rdeephaven/R/table_ops.R

```diff
@@ -0,0 +1,642 @@
+setGeneric(
+  "merge",
+  function(x, y = NULL, ...) {
+    standardGeneric("merge")
+  }
+)
+
+# We need many implementations of 'merge' because of the kind of args it should support
+
+base_merge <- function(x, y, ...) {
+  arg_list <- c(unlist(x), unlist(y), unlist(c(...)))
+  if (length(arg_list) == 1) {
+    return(arg_list[[1]])
+  }
+  unwrapped_arg_list <- lapply(arg_list, strip_s4_wrapping)
+  return(new("TableHandle", .internal_rcpp_object = unwrapped_arg_list[[1]]$merge(unwrapped_arg_list[2:length(unwrapped_arg_list)])))
+}
+
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "TableHandle"),
+  function(x, y = NULL, ...) {
+    return(base_merge(x, y, c(...)))
+  }
+)
+
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "list"),
+  function(x, y = NULL, ...) {
+    return(base_merge(x, y, c(...)))
+  }
+)
+
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "TableHandle", y = "list"),
+  function(x, y, ...) {
+    verify_type("y", y, "TableHandle", "Deephaven TableHandle", FALSE)
+    return(base_merge(x, y, c(...)))
+  }
+)
+
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "list", y = "TableHandle"),
+  function(x, y, ...) {
+    verify_type("x", x, "TableHandle", "Deephaven TableHandle", FALSE)
+    return(base_merge(x, y, c(...)))
+  }
+)
+
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "TableHandle", y = "TableHandle"),
+  function(x, y, ...) {
+    return(base_merge(x, y, c(...)))
+  }
+)
+
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "list", y = "list"),
+  function(x, y, ...) {
+    verify_type("x", x, "TableHandle", "Deephaven TableHandle", FALSE)
+    verify_type("y", y, "TableHandle", "Deephaven TableHandle", FALSE)
+    return(base_merge(x, y, c(...)))
+  }
+)
+
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "NULL", y = "TableHandle"),
+  function(x, y, ...) {
+    return(base_merge(x, y, c(...)))
+  }
+)
+
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "NULL", y = "list"),
+  function(x, y, ...) {
+    verify_type("y", y, "TableHandle", "Deephaven TableHandle", FALSE)
+    return(base_merge(x, y, c(...)))
+  }
+)
+
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "TableHandle", y = "NULL"),
+  function(x, y, ...) {
+    return(base_merge(x, y, c(...)))
+  }
+)
+
+#' @export
+setMethod(
+  "merge",
+  signature = c(x = "list", y = "NULL"),
+  function(x, y, ...) {
+    verify_type("x", x, "TableHandle", "Deephaven TableHandle", FALSE)
+    return(base_merge(x, y, c(...)))
+  }
+)
```

Comment:
It seems like most of our `merge` complexity is from us not limiting support to `list` inputs.  Can you remind me why the various kinds of `TableHandle` inputs seemed like a good idea when we talked?

Reply:
Do we work with multiple list arguments?

Reply:
Now that we see the implementation, would we be better off only supporting a single list arg -- that could contain null, table handles, and lists.  I'm concerned that, given how funky R is, we might miss something.

If we ship just supporting lists, it is easy to add more variants in the future.  If we ship with stuff we don't like, it is harder to take it away.

Reply:
If we're going to restrict support, it should be to both of `merge(t1, t2, t3)` and `merge(c(t1, t2, t3))`. I don't think forcing coercion to a list is desirable if we can easily support `TableHandle` varargs (also easy to type check), and only supporting these two variants reduces the combinatorial mess.

Reply:
`merge(t1,t2,t3)` is not easy to support properly.  It is the source of most of this problem.  Here, any of these can be null.  Then you get a mess.

Reply:
`merge(t1, t2, t3)` is (as far as I understand) easy to support directly, because you can type check varargs as long as they are all of the same type. The source of this problem is trying to support many combinations of TableHandles, lists, and NULLs. I think we can support vararg TableHandles and single list with two methods.

Reply:
Can we support:
1. list
2. varargs of Union[TableHandle,None] (python syntax)?  I think I've seen arrays of types before.